### Ian Chow

### CTA200

### Assignment 2

## Part 1

Write a python function for the function $f(x) = x^3 - x^2 - 1$. Also, write a function for its derivative (you will have to work out $df/dx$ yourself), you can call these functions `f` and `df`.

We know that $f(x) = x^3 - x^2 - 1$. Computing the derivative of $f$, we have:

\begin{equation}
    \frac{df}{dx} = 3x^2 - 2x \nonumber
\end{equation}

$f(x)$ and $df/dx$ are thus implemented in Python as follows:

In [1]:
def f(x):
    """
    finish docs
    """
    return x ** 3 - x ** 2 - 1

def df(x):
    """
    finish docs
    """
    return 3 * x ** 2 - 2 * x

## Part 2:

Write a function `newton(f, df, x0, epsilon=1e-6, max_iter=30)` which performs a Newton Iteration of the function `f` with derivative `df.`

Newton iteration finds the root ($x_n$ such that $f(x_n) = 0$). 

To do this, implement the recursive expression $x_{n + 1} = x_n - \frac{f(x_n)}{f'(x_n)}$ using a loop.

The iteration should stop either when `max_iter` is reached or when $f(x_n) <$ `epsilon`. If the method succeeds, (ie $f\left(x_n\right) <$ `epsilon`), then your function should print `"Found root in <N> iterations"` and should return the value of $x_n$. Otherwise, it should print `"Iteration failed"` and return `None`.

Make sure that your function is documented with Numpy style documentation.

Our function `newton` that implements Newton-Raphson iteration in Python is as follows:

In [2]:
def newton(f, df, x0, epsilon=1e-6, max_iter=30):
    """
    finish docs
    """
    xn = x0  # initialize x_n as x_0, the starting guess for Newton-Raphson iteration
    # perform up to a maximum of max_iter iterations
    for i in range(0, max_iter):
        if abs(f(xn)) < epsilon:  # note that we need to use abs of f(x_n) to find if method has converged to a root:
            print(f'Found root in {i} iterations')  # print number of iterations taken to find the root
            return xn
        xn = xn - f(xn)/df(xn)  # implement the recursive expreession x_{n + 1} = x_n - f(x_n)/f'(x_n) as required
        # iteratively updating x_n based on previous expression
    # if after max_iter iterations it can't find anything, print iteration failed and return None
    print('Iteration failed')
    return None

## Part 3:

Try out your function with the function you defined in part 1. You can experiment with setting $x_0$ differently (show at least two examples of $x_0$ in the notebook). Leave `epsilon` and `max_iter` as the default values specified in part $2$. 

Try reducing `epsilon` to $1e-8$. Does it still work? If so, how many more iterations does it take to converge.

We try computing $x_0$ with setting $x_0 = -1, 0.01, 1$ and $5$:

In [3]:
x0s = [-1, 0.01, 1, 5]  # values of x_0 that we want to test:

# try the newton raphson iteration function using different values of x0:
for x0 in x0s:
    print(f'\nx_0 = {x0}:')
    print(newton(f, df, x0))


x_0 = -1:
Found root in 15 iterations
1.4655712348572754

x_0 = 0.01:
Found root in 19 iterations
1.4655714721100956

x_0 = 1:
Found root in 5 iterations
1.4655713749070918

x_0 = 5:
Found root in 7 iterations
1.4655712402015129


As can be seen, Newton-Raphson iteration converges to the single root $x \sim 1.4656$ for different values of $x_0$. For values far away from the root, it may not converge within the default maxmimum number of iterations since it might be too far away from the root to reach within a certain number of steps. Moreover, for values of $x_0$ very close to $0$ it may also not converge within the maximum number of iterations. Since $f'(0) = 0$, the dividing by `df` step is very sensitive to small changes for $x_0$ close to $0$. This can be seen in our testing, as the $x_0 = 0.01$ takes more iterations than the other three tested.

Now we reduce `epsilon` to $1e-8$ in our function and try the Newton-Raphson iteration function again:

In [4]:
# try newton raphson iteration function using epsilon = 10^-8 this time:
print('epsilon = 10^-8:')

for x0 in x0s:
    print(f'\nx_0 = {x0}:')
    print(newton(f, df, x0, epsilon=1e-8))  # change default value of epsilon this time

epsilon = 10^-8:

x_0 = -1:
Found root in 16 iterations
1.465571231876768

x_0 = 0.01:
Found root in 20 iterations
1.465571231876824

x_0 = 1:
Found root in 6 iterations
1.4655712318767877

x_0 = 5:
Found root in 8 iterations
1.4655712318767682


The function still works as expected, but it now takes more iterations to converge (since $\epsilon$ is smaller). For all four values of $x_0$, the function took $1$ more iteration to converge to the smaller tolerance $\epsilon = 10^{-8}$ compared to $\epsilon = 10^{-6}$.